### PR TITLE
Fix #1181 #1184 - Support adding custom filters

### LINKED ISSUE #1181: Need a solution to add few custom filters before and after the default micro gateway handlers
### Describe your problem(s)
Need a solution to add few custom filters before and after the default micro gateway handlers
There are cases we need to add filters before the API Schema Validation too.

### How will you implement it

---

### LINKED ISSUE #1184: Cannot add extensions - file does not copy from extensions to the target
### Description:
Cannot add extensions - file does not copy from extensions to the target
Right now we put extensions in interceptor folder and copy to the target

### Environment details (with versions):
- OS:
- Client:
- Env (Docker/K8s):

---

### PR DESCRIPTION
**Pull request by Rajith90**
### Purpose
With this fix specifying custom filters are supported. Developer can define custom filters and the position of that filter[1] in the filter chain in the deployment-config.toml, and then
add that filter inside the extension folder of the project.

[1]  - 
```toml
[[filters]]
    name = "CustomFilter"
    position = 5
[[filters]]
    name = "CustomFilterNew"
    position = 2
```

### Issues

Fixes #1181 
Fixes #1184 

### Automation tests
 - Unit tests added: No
 - Integration tests added: No

### Tested environments

Not Tested

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: components/micro-gateway-cli/src/main/resources/templates/listeners.mustache [hunk 1]
```diff
@@ -133,7 +134,40 @@ public function getFilters() returns (http:RequestFilter|http:ResponseFilter) []
...
+    customFilters[{{@index}}] = {{variableName}};
+    allFilters.push({{variableName}}) ;
+{{/containerConfig.filters}}
+
+    int customFilterIndex = 0;
+    int authFilterPosition = gateway:getAuthFilterPosition();
+
+    foreach var filterIndex in customFilterIndexes {
+        shiftArray(allFilters, filterIndex, customFilters[customFilterIndex]);
+        if(filterIndex <= authFilterPosition) {
```
Comment (praminda) [inline components/micro-gateway-cli/src/main/resources/templates/listeners.mustache:155]:
```suggestion
        if (filterIndex <= authFilterPosition) {
```

File: components/micro-gateway-cli/src/main/resources/templates/listeners.mustache [hunk 2]
```diff
@@ -133,7 +134,40 @@ public function getFilters() returns (http:RequestFilter|http:ResponseFilter) []
...
+        }
+        customFilterIndex = customFilterIndex +1;
+    }
+
+    gateway:setAuthFilterPosition(authFilterPosition);
+    log:printDebug(allFilters.toString());
+    return allFilters;
+}
+
+function shiftArray((http:RequestFilter|http:ResponseFilter)[] allFilters, int position, (http:RequestFilter|http:ResponseFilter) customFilter){
```
Comment (praminda) [inline components/micro-gateway-cli/src/main/resources/templates/listeners.mustache:166]:
```suggestion
function shiftArray((http:RequestFilter|http:ResponseFilter)[] allFilters, int position, (http:RequestFilter|http:ResponseFilter) customFilter) {
```

File: components/micro-gateway-cli/src/main/resources/templates/listeners.mustache [hunk 3]
```diff
@@ -133,7 +134,40 @@ public function getFilters() returns (http:RequestFilter|http:ResponseFilter) []
...
+    int n = allFilters.length();
```
Comment (AmaliMatharaarachchi) [inline components/micro-gateway-cli/src/main/resources/templates/listeners.mustache:167]: It's better to provide meaningful variables for n and i

File: components/micro-gateway-core/src/main/ballerina/src/gateway/constants/constants.bal [hunk 4]
```diff
@@ -426,3 +426,5 @@ public const string API_PUBLISHER = "api_publisher";
 public const string API_NAME = "api_name";
 public const string MATCHING_RESOURCE = "matching_resource";
 
+
```
[the remark this answers is not in the document]
Comment (Rajith90) [reply components/micro-gateway-core/src/main/ballerina/src/gateway/constants/constants.bal:429]: This separation is required to define different set of constants. Each separated block defines set of constants related to a single feature or capability

**Pull request merged**